// @the-/error/shim/esm/index.js
// Code generated by coz. DO NOT EDIT.

/**
 * @module @the-/error
 * @version 15.4.0
 * @description Error for the-framework
 * @typicalname error
 * @license MIT
 */
'use strict';

import TheConflictError_ from './TheConflictError';
import TheError_ from './TheError';
import TheExpiredError_ from './TheExpiredError';
import TheForbiddenError_ from './TheForbiddenError';
import TheGoneError_ from './TheGoneError';
import TheInvalidParameterError_ from './TheInvalidParameterError';
import TheNotAcceptableError_ from './TheNotAcceptableError';
import TheNotFoundError_ from './TheNotFoundError';
import TheRequestTimeoutError_ from './TheRequestTimeoutError';
import TheServerRottenError_ from './TheServerRottenError';
import TheUnauthorizedError_ from './TheUnauthorizedError';
import create_ from './create';
import define_ from './define'; // `module.exports` overrides these `exports.*`, but still needs them for lebab (https://github.com/lebab/lebab)

export { TheConflictError_ as TheConflictError };
export { TheError_ as TheError };
export { TheExpiredError_ as TheExpiredError };
export { TheForbiddenError_ as TheForbiddenError };
export { TheGoneError_ as TheGoneError };
export { TheInvalidParameterError_ as TheInvalidParameterError };
export { TheNotAcceptableError_ as TheNotAcceptableError };
export { TheNotFoundError_ as TheNotFoundError };
export { TheRequestTimeoutError_ as TheRequestTimeoutError };
export { TheServerRottenError_ as TheServerRottenError };
export { TheUnauthorizedError_ as TheUnauthorizedError };
export { create_ as create };
export { define_ as define };
export default {
  TheConflictError: TheConflictError_,
  TheError: TheError_,
  TheExpiredError: TheExpiredError_,
  TheForbiddenError: TheForbiddenError_,
  TheGoneError: TheGoneError_,
  TheInvalidParameterError: TheInvalidParameterError_,
  TheNotAcceptableError: TheNotAcceptableError_,
  TheNotFoundError: TheNotFoundError_,
  TheRequestTimeoutError: TheRequestTimeoutError_,
  TheServerRottenError: TheServerRottenError_,
  TheUnauthorizedError: TheUnauthorizedError_,
  create: create_,
  define: define_
};
