// @the-/error/shim/index.js
// Code generated by coz. DO NOT EDIT.

/**
 * @module @the-/error
 * @version 15.4.0
 * @description Error for the-framework
 * @typicalname error
 * @license MIT
 */
'use strict';

var TheConflictError_ = require('./TheConflictError');

var TheError_ = require('./TheError');

var TheExpiredError_ = require('./TheExpiredError');

var TheForbiddenError_ = require('./TheForbiddenError');

var TheGoneError_ = require('./TheGoneError');

var TheInvalidParameterError_ = require('./TheInvalidParameterError');

var TheNotAcceptableError_ = require('./TheNotAcceptableError');

var TheNotFoundError_ = require('./TheNotFoundError');

var TheRequestTimeoutError_ = require('./TheRequestTimeoutError');

var TheServerRottenError_ = require('./TheServerRottenError');

var TheUnauthorizedError_ = require('./TheUnauthorizedError');

var create_ = require('./create');

var define_ = require('./define'); // `module.exports` overrides these `exports.*`, but still needs them for lebab (https://github.com/lebab/lebab)


exports.TheConflictError = TheConflictError_;
exports.TheError = TheError_;
exports.TheExpiredError = TheExpiredError_;
exports.TheForbiddenError = TheForbiddenError_;
exports.TheGoneError = TheGoneError_;
exports.TheInvalidParameterError = TheInvalidParameterError_;
exports.TheNotAcceptableError = TheNotAcceptableError_;
exports.TheNotFoundError = TheNotFoundError_;
exports.TheRequestTimeoutError = TheRequestTimeoutError_;
exports.TheServerRottenError = TheServerRottenError_;
exports.TheUnauthorizedError = TheUnauthorizedError_;
exports.create = create_;
exports.define = define_;
module.exports = {
  TheConflictError: TheConflictError_,
  TheError: TheError_,
  TheExpiredError: TheExpiredError_,
  TheForbiddenError: TheForbiddenError_,
  TheGoneError: TheGoneError_,
  TheInvalidParameterError: TheInvalidParameterError_,
  TheNotAcceptableError: TheNotAcceptableError_,
  TheNotFoundError: TheNotFoundError_,
  TheRequestTimeoutError: TheRequestTimeoutError_,
  TheServerRottenError: TheServerRottenError_,
  TheUnauthorizedError: TheUnauthorizedError_,
  create: create_,
  define: define_
};
